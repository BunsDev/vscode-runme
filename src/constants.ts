/**
 * Note: this file is used within Node.js and Browser environment.
 * Only export cross compatible objects here.
 */

export enum OutputType {
  shell = 'stateful.runme/shell-stdout',
  vercel = 'stateful.runme/vercel-stdout',
  html = 'stateful.runme/html-stdout',
  script = 'stateful.runme/script-stdout',
  error = 'error'
}

export const CONFIGURATION_SHELL_DEFAULTS = {
  interactive: true,
  closeTerminalOnSuccess: true
} as const
