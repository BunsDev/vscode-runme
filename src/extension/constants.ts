export const DEFAULT_ENV = {
  RUNME_TASK: 'true',
  PATH: process.env.PATH || ''
}
export const ENV_STORE = new Map<string, string>(
  Object.entries(DEFAULT_ENV)
)
