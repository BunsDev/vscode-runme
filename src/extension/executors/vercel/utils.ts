import fs from 'node:fs/promises'
import path from 'node:path'

import xdg from 'xdg-app-paths'

export async function getConfigFilePath () {
  return path.join(
    `${xdg('com.vercel.cli').dataDirs()[0]}.cli`, 'auth.json')
}

export async function getAuthToken () {
  const authFilePath = await getConfigFilePath()

  try {
    const canRead = await fs.access(authFilePath).then(() => true, () => false)
    if (canRead) {
      return JSON.parse(
        (await fs.readFile(authFilePath, 'utf-8')).toString()
      ).token as string
    }
  } catch (err: any) {
    return
  }
}
