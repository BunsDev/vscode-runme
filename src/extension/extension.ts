import * as vscode from "vscode"
import {
  Kernel,
  Serializer,
  ThumbsDownProvider,
  ThumbsUpProvider,
} from "./providers"

export function activate(context: vscode.ExtensionContext) {
  context.subscriptions.push(
    vscode.workspace.registerNotebookSerializer("runme", new Serializer(context), {
      transientOutputs: true,
      transientCellMetadata: {
        inputCollapsed: true,
        outputCollapsed: true,
      },
    }),
    vscode.notebooks.registerNotebookCellStatusBarItemProvider(
      "runme",
      new ThumbsUpProvider()
    ),
    vscode.notebooks.registerNotebookCellStatusBarItemProvider(
      "runme",
      new ThumbsDownProvider()
    )
  )
  context.subscriptions.push(new Kernel())
}

// This method is called when your extension is deactivated
export function deactivate() { }
