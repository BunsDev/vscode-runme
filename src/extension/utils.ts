import vscode from 'vscode'

export function isInteractiveTask (cell: vscode.NotebookCell) {
  const config = vscode.workspace.getConfiguration('runme')

  /**
   * if cell is marked as interactive (default: not set or set to 'true')
   */
  if (
    typeof cell.metadata.attributes?.interactive === 'undefined' ||
    cell.metadata.attributes.interactive === 'true'
  ) {
    return true
  }

  /**
   * if it is set within the settings (default: true)
   */
  if (config.get('shell.interactive')) {
    return true
  }

  return false
}

export function getTerminalByCell (cell: vscode.NotebookCell) {
  return vscode.window.terminals.find((t) => {
    const taskEnv = (t.creationOptions as vscode.TerminalOptions).env || {}
    return taskEnv.RUNME_ID === `${cell.document.fileName}:${cell.index}`
  })
}
