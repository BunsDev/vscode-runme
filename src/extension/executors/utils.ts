import { NotebookCellOutput, NotebookCellExecution, NotebookCellOutputItem } from 'vscode'

import { OutputType } from '../../constants'
import type { CellOutput } from '../../types'

export function renderError (exec: NotebookCellExecution, output: string) {
  return exec.replaceOutput(new NotebookCellOutput([
    NotebookCellOutputItem.json(
      <CellOutput<OutputType.error>>{ type: 'error', output },
      OutputType.deno
    )
  ]))
}
