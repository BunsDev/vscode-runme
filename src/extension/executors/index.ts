import { sh, bash } from './task'
import { vercel } from './vercel'
import { html } from './html'

export default { sh, bash, vercel, html }
