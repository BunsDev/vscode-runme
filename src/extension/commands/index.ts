import path from 'node:path'

import {
  NotebookCell, Uri, window, env, NotebookDocument, TextDocument, ViewColumn,
  workspace, NotebookData, commands, NotebookCellData, NotebookCellKind
} from 'vscode'

import { CliProvider } from '../provider/cli'
import { getMetadata, getTerminalByCell } from '../utils'

function showWarningMessage () {
  return window.showWarningMessage('Couldn\'t find terminal! Was it already closed?')
}

export function openTerminal (cell: NotebookCell) {
  const terminal = getTerminalByCell(cell)
  if (!terminal) {
    return showWarningMessage()
  }
  return terminal.show()
}

export function copyCellToClipboard (cell: NotebookCell) {
  env.clipboard.writeText(cell.document.getText())
  return window.showInformationMessage('Copied cell to clipboard!')
}

export function stopBackgroundTask (cell: NotebookCell) {
  const terminal = getTerminalByCell(cell)
  if (!terminal) {
    return showWarningMessage()
  }
  terminal.dispose()
  return window.showInformationMessage(`${terminal?.name} task terminated!`)
}

export async function runCLICommand (cell: NotebookCell) {
  if (!await CliProvider.isCliInstalled()) {
    return window.showInformationMessage(
      'Runme CLI is not installed. Do you want to download it?',
      'Download now'
    ).then((openBrowser) => openBrowser && env.openExternal(
      Uri.parse('https://github.com/stateful/runme/releases')
    ))
  }
  const metadata = getMetadata(cell)
  const cliName: string = (metadata?.['runme.dev/name'] || '').trim()
  const term = window.createTerminal(`CLI: ${cliName}`)
  term.show(false)
  term.sendText(`runme run ${cliName} --chdir="${path.dirname(cell.document.uri.fsPath)}"`)
}
export function openAsRunmeNotebook (doc: NotebookDocument) {
  window.showNotebookDocument(doc, {
    viewColumn: ViewColumn.Beside
  })
}

export function openSplitViewAsMarkdownText (doc: TextDocument) {
  window.showTextDocument(doc, {
    viewColumn: ViewColumn.Beside
  })
}

export async function createNewRunmeNotebook () {
  const newNotebook = await workspace.openNotebookDocument('runme', new NotebookData([
    new NotebookCellData(
      NotebookCellKind.Markup,
      '# Runme Notebook\n\nStart writing here...',
      'markdown'
    ),
    new NotebookCellData(
      NotebookCellKind.Code,
      'echo "Hello World"',
      'bash'
    ),
    new NotebookCellData(
      NotebookCellKind.Markup,
      '_Find more documentation on writing Runme markdown on [runme.dev](https://www.runme.dev/docs/intro)!_',
      'markdown'
    )
  ]))
  await commands.executeCommand('vscode.openWith', newNotebook.uri, 'runme')
}
