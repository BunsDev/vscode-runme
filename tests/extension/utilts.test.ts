import vscode from 'vscode'
import { expect, vi, test, beforeAll, afterAll } from 'vitest'

import { getExecutionProperty, getTerminalByCell, populateEnvVar, resetEnv, getKey } from '../../src/extension/utils'
import { ENV_STORE, DEFAULT_ENV } from '../../src/extension/constants'

vi.mock('vscode', () => ({
  default: {
    window: {
      terminals: [
        { creationOptions: { env: {} } },
        { creationOptions: { env: { RUNME_ID: 'foobar:123' } } }
      ]
    },
    workspace: {
      getConfiguration: vi.fn()
    }
  }
}))

const PATH = process.env.PATH
beforeAll(() => {
  DEFAULT_ENV.PATH = '/usr/bin'
  ENV_STORE.delete('PATH')
})
afterAll(() => { process.env.PATH = PATH })

test('isInteractive', () => {
  // when set to false in configutaration
  vi.mocked(vscode.workspace.getConfiguration).mockReturnValue({ get: vi.fn().mockReturnValue(false) } as any)
  expect(getExecutionProperty('interactive', { metadata: {} } as any)).toBe(false)
  expect(getExecutionProperty('interactive', { metadata: { attributes: {} } } as any)).toBe(false)
  expect(getExecutionProperty('interactive', { metadata: { attributes: { interactive: 'true' } } } as any)).toBe(true)

  vi.mocked(vscode.workspace.getConfiguration).mockReturnValue({ get: vi.fn().mockReturnValue(true) } as any)
  expect(getExecutionProperty('interactive', { metadata: {} } as any)).toBe(true)
  expect(getExecutionProperty('interactive', { metadata: { attributes: {} } } as any)).toBe(true)
})

test('getTerminalByCell', () => {
  expect(getTerminalByCell({ document: { fileName: 'foo' }, index: 42} as any))
    .toBe(undefined)
  expect(getTerminalByCell({ document: { fileName: 'foobar' }, index: 123} as any))
    .not.toBe(undefined)
})

test('populateEnvVar', () => {
  expect(populateEnvVar(
    'export PATH="/foo/$BAR/$LOO:$PATH:/$FOO"',
    { PATH: '/usr/bin', FOO: 'foo', BAR: 'bar' }
  )).toBe('export PATH="/foo/bar/:/usr/bin:/foo"')
})

test('resetEnv', () => {
  ENV_STORE.set('foo', 'bar')
  expect(ENV_STORE).toMatchSnapshot()
  resetEnv()
  expect(ENV_STORE).toMatchSnapshot()
})

test('getKey', () => {
  expect(getKey({
    getText: vi.fn().mockReturnValue('foobar'),
    languageId: 'barfoo'
  } as any)).toBe('barfoo')
  expect(getKey({
    getText: vi.fn().mockReturnValue('deployctl deploy foobar'),
    languageId: 'something else'
  } as any)).toBe('deno')
})
