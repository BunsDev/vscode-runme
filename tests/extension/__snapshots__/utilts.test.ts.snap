// Vitest Snapshot v1

exports[`resetEnv 1`] = `
Map {
  "RUNME_TASK" => "true",
  "foo" => "bar",
}
`;

exports[`resetEnv 2`] = `
Map {
  "RUNME_TASK" => "true",
  "PATH" => "/usr/bin",
}
`;
